### DISCUSSION TITLE
RPM on PyPI

### DISCUSSION THREAD

**Discussion by nforro** [rpm-software-management/rpm#2361] [General]
There is a [dummy RPM package](https://pypi.org/project/rpm/) on PyPI that intentionally fails to install.
I'm wondering if a shim module similar to [selinux](https://pypi.org/project/selinux/) wouldn't be beneficial. It would replace itself with system RPM module after import or raise an error stating that one needs to install RPM python bindings on their system. What do you think?

**Comment by TomasTomecek** [comment]
This would be an improvement to the existing workflow where folks outside of rpm-based distros can't even install the tooling.

**Comment by pmatilai** [comment]
We don't know anything about that dummy package, but in general we don'tr really want rpm on Pypi, it's not well suited to that.

**Reply by nforro** [reply]
> in general we don'tr really want rpm on Pypi, it's not well suited to that

And would you mind if we (the @packit team) created and maintained such shim module?

**Reply by pmatilai** [reply]
I'm not sure I understand the case for this shim, but it doesn't sound particularly harmful (whereas trying to make actual rpm bindings available in PyPi is). So I guess that means "no we don't mind".

**Reply by nforro** [reply]
Thank you. The way I see it, the benefits of this shim module are:
- we can declare dependency on `rpm` while keeping our projects installable with pip, which also allows us to take advantage of [automatic dependency generators](https://docs.fedoraproject.org/en-US/packaging-guidelines/Python/#Automatically-generated-dependencies) without any hacks
- thanks to the shim, our projects can be installed and run in virtualenvs, using system `rpm` module

**Comment by TomasTomecek** [comment]
I can see it's Nick Coghlan who maintains that namespace, I believe we should reach out to him and discuss this. Thank you for your response, Panu!

**Comment by hroncok** [comment]
See also https://pypi.org/project/rpm-py-installer/

**Reply by TomasTomecek** [reply]
Actually rpm-py-installer triggered this conversation: https://github.com/packit/specfile/issues/187

**Comment by hroncok** [comment]
Note that if this is ready https://github.com/rpm-software-management/rpm/issues/2345 -- the rpm shim should be able to copy/link the _rpm extension module from the python3-rpm RPM package to virtual environments using different (well, probably just newer) Python versions.

**Reply by nforro** [reply]
That would be great and it would pretty much eliminate the need for rpm-py-installer, at least on rpm-based systems.

**Comment by thrix** [comment]
@nforro @hroncok @TomasTomecek thanks for the discussion, so this is blocked on  #2345 ?

We hit this again with new pip, and need to stick to pip < 23.1, where even the old does not install anymore ;(

https://gitlab.com/testing-farm/gluetool-modules/-/merge_requests/438/diffs

**Reply by nforro** [reply]
No, `specfile` now depends on the shim module rather than `rpm-py-installer`.

**Reply by thrix** [reply]
maybe to late for me, I have issues with `rpm-py-installer`, I thought then I need to go back to cry to https://github.com/junaruga/rpm-py-installer/issues/260 :)

**Reply by nforro** [reply]
You need it for koji? Couldn't they also switch to the [shim module](https://github.com/packit/rpm-shim)?

**Reply by thrix** [reply]
nope, our own tooling, gluetool, let us try shim, I was not aware it exists
